python/tvm/relay/op/strategy/cuda.py

```diff
@@ -222,12 +218,11 @@ def conv2d_strategy_cuda(attrs, inputs, out_type, target):
 
             # register auto-scheduler implementations
             if judge_winograd_auto_scheduler:
-                strategy.add_auto_scheduler(
-                    wrap_compute_conv2d(topi.nn.conv2d_winograd_nhwc), name="conv2d_nhwc.winograd"
-                )
-            else:
-                strategy.add_auto_scheduler(
-                    wrap_compute_conv2d(topi.nn.conv2d_nhwc), name="conv2d_nhwc"
+                strategy.add_implementation(
+                    wrap_compute_conv2d(topi.nn.conv2d_winograd_nhwc),
+                    wrap_topi_schedule(tvm.te.create_schedule),
+                    name="conv2d_nhwc.winograd",
+                    plevel=10,
```

Comment:
Both the plevel for `conv2d_nhwc.cuda` and `conv2d_nhwc.winograd` are 10.
Which one will auto-scheduler pick? Which one will autotvm pick?

Reply:
I was thinking about this too. The current semantic is when extracting AutoTVM tasks, only the direct one will be extracted because Winograd one isn't registered as an AutoTVM task, so does the case of applying AutoTVM tuning logs. So we can probably just make this plevel higher (say, 12).

The only issue is when no tuning logs are given, the Winograd one will be selected if judge_Winograd is True. In this case we need to make sure Winograd fallback is better than the direct.

What do you think?

Reply:
Autotvm should never select this winograd implementation because there is no schedule templates for it.
We should have a way to specify this. But as you said, a single plevel cannot work for both autotvm and auto-scheduler.

Reply:
Hmm you're right. We can never pick Winograd when auto_scheduler is not enabled. Then how about we check PassContext and only add this Winograd when use_auto_scheduler is True? In long term we will need a better mechanism to select the implementation anyways.

Reply:
Sounds good